server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -337,13 +337,11 @@ void buildDiffAndSerializeStates() {
                     // publication to local node bypasses any serialization
                     continue;
                 }
+                TransportVersion version = transportService.getConnection(node).getTransportVersion();
```

Comment:
A lot of tests are failing because the node doesn't necessarily have a connection yet at this point...

Reply:
In production, existing nodes should be connected due to `NodeConnectionsService`, and we should already have connected to new nodes here:

https://github.com/elastic/elasticsearch/blob/405799652508e696a23bde9a907f502ca3647248/server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java#L596

We shouldn't throw a NPE here, we should treat it as if the node is disconnected (i.e. pass a `NodeNotConnectedException` to the relevant listener)

Reply:
At this point, the only listener available is over in `Coordinator.publish`, being caught by the `Exception` handler in that method. This is causing the whole state publish to fail, which is breaking the system (`CoordinatorTests.testLeaderDisconnectionWithDisconnectEventDetectedQuickly` specifically)

Reply:
Could we do something here like use a fallback/default transport version if the connection is not yet established, as the version here is just used for serialisation? So if the node isn't connected, we don't need to worry about it reading anything.

Reply:
If we don't have a connection by this point then there's no reason to believe we will have a connection by the point we want to send something either. You can just skip disconnected nodes.

Reply:
> (i.e. pass a `NodeNotConnectedException` to the relevant listener)

Sorry, I didn't mean to suggest that here. This advice only applies to the other call sites.

Reply:
Great